core/shared/src/main/scala/fs2/Pull.scala

```diff
@@ -48,20 +123,85 @@ import Pull._
+    *
+    * This operation does not modify resource scope boundaries. The generated `post` 
+    * pull starts compilation on the same scope on which the compilation of this 
+    * pull (the prefix) ends. The composed pull will end on whatever scope the 
+    * compilation of the `post` pull does.
+    */
+  def flatMap[F2[x] >: F[x], O2 >: O, R2](kont: R => Pull[F2, O2, R2]): Pull[F2, O2, R2] =
+    new Bind[F2, O2, R, R2](this) {
+      def cont(e: Terminal[R]): Pull[F2, O2, R2] =
+        e match {
+          case Succeeded(r) =>
+            try kont(r)
+            catch { case NonFatal(e) => Fail(e) }
+          case res @ Interrupted(_, _) => res
+          case res @ Fail(_)           => res
+        }
+    }
 
-  /** Returns a pull with the result wrapped in `Right`, or an error wrapped in `Left` if the pull has failed. */
-  def attempt: Pull[F, O, Either[Throwable, R]] =
-    map(r => Right(r)).handleErrorWith(t => Succeeded(Left(t)))
+  /** Lazily appends the given `post` pull, to be run after `this` pull.
+    *
+    * - If `this` pull completes successfully, the `post` argument is evaluated to build 
+    *   the pull (this is a lazy parameter), and the compilation starts running the `post` 
+    *   pull. The result of the `post` pull is the result of the composed pull.
+    *   Any carry-over value from `this` pull is discarded.
+    *
+    * - If the compilation of `this` pull is failed or interrupted, then the `post` argument
+    *   is not evaluated and the composed pull returns with the failure or interruption.
+    *
+    * Note that, because the `post` argument is lazy, this method can be used 
+    * to build lazy pulls, in which latest parts of the Pulls are not allocated
+    * into memory until after the first parts have been successfully compiled.
+    *
+    * In both cases, the effectful actions and outputs of the appended pull 
+    * consists of those outputs and actions from the first pull, followed by
+    * those from the `post` pull, in the same order as they would come out of each pull.
+    *
+    * This operation does not add or remove any resource scope boundaries. 
+    * The `post` pull will run on the same scope on which the compilation of 
+    * this pull ends. The composed pull will end on whatever scope the `post` 
+    * pull ended on.
+    */
+  def >>[F2[x] >: F[x], O2 >: O, S](post: => Pull[F2, O2, S]): Pull[F2, O2, S] =
+    new Bind[F2, O2, R, S](this) {
+      def cont(r: Terminal[R]): Pull[F2, O2, S] =
+        r match {
+          case _: Succeeded[_] => post
+          case r: Interrupted  => r
+          case r: Fail         => r
+        }
+    }
 
   /** Interpret this `Pull` to produce a `Stream`, introducing a scope.
     *
-    * May only be called on pulls which return a `Unit` result type. Use `p.void.stream` to explicitly
-    * ignore the result type of the pull.
+    * The transformation to `Stream` wraps the given `this` pull to be run in 
```

Comment:
I'd remove this paragraph or otherwise reword to "A scope is introduced to ensure any resources acquired in this pull are released in a timely fashion, even when the resulting stream is not run until completion."